File: leetcode/algorithms/345_reverse_vowels_of_a_string/main.js
/**
 * Two Pointers
 *
 * Complexities:
 *   N - The size of `s`
 *   - Time Complexities: O(N)
 *   - Space Complexities: O(N)
 */
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function (s) {
  const vowels = ["A", "E", "I", "O", "U", "a", "e", "i", "o", "u"];
  const vowelStack = [];
  let result = "";

  for (const c of s) {
    if (vowels.includes(c)) {
      vowelStack.push(c);
    }
  }

  for (let i = 0; i < s.length; i++) {
    if (vowels.includes(s[i])) {
      result += vowelStack.pop();
    } else {
      result += s[i];
    }
  }

  return result;
};


// Solution
/**
 * Solution 1
 *
 * Two Pointers
 *
 * Complexities:
 *   N - The size of `s`
 *   - Time Complexities: O(N)
 *   - Space Complexities: O(N)
 */
/**
 * @param {string} s
 * @return {string}
 */
var solution1 = function (s) {
  const vowels = new Set(["a", "e", "i", "o", "u", "A", "E", "I", "O", "U"]);
  const arr = s.split("");
  let left = 0;
  let right = arr.length - 1;

  while (left < right) {
    while (left < right && !vowels.has(arr[left])) {
      left++;
    }
    while (left < right && !vowels.has(arr[right])) {
      right--;
    }

    if (left < right) {
      [arr[left], arr[right]] = [arr[right], arr[left]];
      left++;
      right--;
    }
  }

  return arr.join("");
};

/**
 * Solution 2
 *
 * Two Pointers
 *
 * Complexities:
 *   N - The size of `s`
 *   - Time Complexities: O(N)
 *   - Space Complexities: O(N)
 */
/**
 * @param {string} s
 * @return {string}
 */
const isVowel = (c) => "aeiou".includes(c.toLowerCase());

var solution2 = function (s) {
  const vowels = new Set(["a", "e", "i", "o", "u", "A", "E", "I", "O", "U"]);
  const arr = s.split("");
  let left = 0;
  let right = arr.length - 1;

  while (left < right) {
    while (left < right && !vowels.has(arr[left])) {
      left++;
    }
    while (left < right && !vowels.has(arr[right])) {
      right--;
    }

    if (left < right) {
      [arr[left], arr[right]] = [arr[right], arr[left]];
      left++;
      right--;
    }
  }

  return arr.join("");
};

/**
 * Solution 3
 *
 * Two Pointers
 *
 * Complexities:
 *   N - The size of `s`
 *   - Time Complexities: O(N)
 *   - Space Complexities: O(N)
 */
/**
 * @param {string} s
 * @return {string}
 */
var solution3 = function (s) {
  const vowels = new Set("aeiouAEIOU");
  const found = [];

  for (const c of s) {
    if (vowels.has(c)) {
      found.push(c);
    }
  }

  let idx = found.length - 1;
  return s.replace(/[aeiouAEIOU]/g, () => found[idx--]);
};

File: leetcode/algorithms/345_reverse_vowels_of_a_string/main.py
class ReverseVowelsOfAString:
    """
    # Two Pointers
    #
    # Complexities:
    #   N - The size of `s`
    #   - Time Complexities: O(N)
    #   - Space Complexities: O(N)
    """
    def reverseVowels(self, s: str) -> str:
        result = ""
        vowels = ["A", "E", "I", "O", "U", "a", "e", "i", "o", "u"]
        vowelStack = []

        for c in s:
            if c in vowels:
                vowelStack.append(c)

        for i in range(len(s)):
            if s[i] in vowels:
                result += vowelStack.pop()
            else:
                result += s[i]

        return result


    # Solution
    """
    # Solution 1
    #
    # Two Pointers
    #
    # Complexities:
    #   N - The size of `s`
    #   - Time Complexities: O(N)
    #   - Space Complexities: O(N)
    """
    def solution1(self, s: str) -> str:
        vowels = set('aeiouAEIOU')
        chars = list(s)
        left, right = 0, len(chars) - 1

        while left < right:
            while left < right and chars[left] not in vowels:
                left += 1
            while left < right and chars[right] not in vowels:
                right -= 1

            chars[left], chars[right] = chars[right], chars[left]
            left += 1
            right -= 1

        return ''.join(chars)

    """
    # Solution 2
    #
    # Complexities:
    #   N - The size of `s`
    #   - Time Complexities: O(N)
    #   - Space Complexities: O(N)
    """
    def solution2(self, s: str) -> str:
        vowels = set('aeiouAEIOU')
        found = [c for c in s if c in vowels]
        chars = list(s)

        for i, c in enumerate(chars):
            if c in vowels:
                chars[i] = found.pop()

        return ''.join(chars)
